internals/overlord/tlsstate/manager.go

```diff
@@ -0,0 +1,511 @@
+	}
+	var pemBuffer bytes.Buffer
+	if err := pem.Encode(&pemBuffer, pemBlock); err != nil {
+		return nil, fmt.Errorf("cannot convert certificate to PEM: %w", err)
+	}
+	return pemBuffer.Bytes(), nil
+}
+
+// pemKey returns a PKCS8 encoded private key in PEM format.
+func (t TLSKeyPair) pemKey() ([]byte, error) {
+	privateBytes, err := x509.MarshalPKCS8PrivateKey(t.privateKey)
+	if err != nil {
+		return nil, err
+	}
+	pemPrivateBlock := &pem.Block{
+		Type:  "PRIVATE KEY",
+		Bytes: privateBytes,
+	}
+	var pemBuffer bytes.Buffer
+	if err := pem.Encode(&pemBuffer, pemPrivateBlock); err != nil {
+		return nil, fmt.Errorf("cannot convert key to PEM: %w", err)
+	}
+	return pemBuffer.Bytes(), nil
+}
+
+// toPEMBytes converts the private key and x509 certificate into a byte slice
+// that contains two PEM blocks, ready for secure storage on disk.
+func (t TLSKeyPair) toPEMBytes() ([]byte, error) {
+	privateBytes, err := x509.MarshalPKCS8PrivateKey(t.privateKey)
+	if err != nil {
+		return nil, err
+	}
+	pemPrivateBlock := &pem.Block{
+		Type:  "PRIVATE KEY",
+		Bytes: privateBytes,
+	}
+	pemCertificateBlock := &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: t.Certificate.Raw,
+	}
+	var pemBuffer bytes.Buffer
+	if err := pem.Encode(&pemBuffer, pemPrivateBlock); err != nil {
+		return nil, fmt.Errorf("cannot convert key to PEM: %w", err)
+	}
+	if err := pem.Encode(&pemBuffer, pemCertificateBlock); err != nil {
+		return nil, fmt.Errorf("cannot convert x509 certificate to PEM: %w", err)
+	}
+	return pemBuffer.Bytes(), nil
+}
+
+// fromPEMFile loads a PKCS8 private key and x509 certificate
+// from a PEM formatted file.
+func (t *TLSKeyPair) fromPEMFile(path string) error {
+	pemData, err := os.ReadFile(path)
+	if err != nil {
+		return err
+	}
+	for {
+		var block *pem.Block
+		block, pemData = pem.Decode(pemData)
+		if block == nil {
+			break
+		}
+		switch block.Type {
+		case "PRIVATE KEY":
+			t.privateKey, err = x509.ParsePKCS8PrivateKey(block.Bytes)
+			if err != nil {
+				return err
+			}
+		case "CERTIFICATE":
+			t.Certificate, err = x509.ParseCertificate(block.Bytes)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	if t.privateKey == nil || t.Certificate == nil {
+		return fmt.Errorf("PEM file does not contain the expected keypair")
+	}
+	return nil
+}
+
+// isExpired returns true if the X509 certificate has expired.
+func (t TLSKeyPair) isExpired() bool {
+	now := systemTime()
+	if now.Before(t.Certificate.NotBefore) || now.After(t.Certificate.NotAfter) {
+		return true
+	}
+	return false
+}
+
+// addFingerprint adds the base32 string encoded fingerprint of the public certificate
+// so we can access it later as a cached value. No fingerprint will be added if
+// this method is called before a certificate was added.
+func (t *TLSKeyPair) addFingerprint() {
+	if t.Certificate != nil {
+		hashBytes := sha512.Sum384(t.Certificate.Raw)
+		t.fingerprint = base32.StdEncoding.WithPadding(base32.NoPadding).EncodeToString(hashBytes[:])
+	}
+}
```

Comment:
This function doesn't need a keypair as input, only the certificate.

```suggestion
func fingerprint(cert *x509.Certificate) string {
  hashBytes := sha512.Sum384(cert.Raw)
  return base32.StdEncoding.WithPadding(base32.NoPadding).EncodeToString(hashBytes[:])
}
```

And I think you could just inline this code into the function that creates a new keypair. It doesn't make sense to call this function anywhere else anyway.